### DISCUSSION TITLE
PMTiles overlay renderer in Swift that works with MapKit

### DISCUSSION THREAD

**Discussion by hardeverick** [protomaps/PMTiles#486] [Q&A] [answered]
Hi all,

We're trying to show overlay of custom vector data on top of a MapKit map. We have an app that uses MapKit for it's maps. At the moment we cannot switch to something like MapLibre GL Native since we currently depend on MapKit. MapKit has the `MKTileOverlay` class which overlays bitmap data (not vector). Is there a way to render the vector tiles on device and then provide them to MapKit? I'm also open to any other solution.
Would porting https://github.com/protomaps/protomaps-leaflet be worthwhile? It seems like that's doing the same thing I want to do but for leaflet.

Thanks,

Edit: We would also like to avoid running a separate server if at all possible

**Comment by bdon** [comment, marked as answer]
I don't know of any MVT vector tile renderer that works with mapkit.

**Reply by hardeverick** [reply]
@bdon Thanks for the reply. I've started the process of porting over the pmtiles parser/fetcher and once I have the MVT data I will then port the rendering code from `protomaps-leaflet` to render it. So I'll probably have some questions in the future.

**Reply by bdon** [reply]
I don't recommend porting over that rendering code as it is quite specific to the leaflet rendering model and the web canvas API. You should use iOS native hardware accelerated drawing and text rendering APIs.

**Reply by hardeverick** [reply]
I see. I ported the v3 pmtile fetching code and I'm getting valid vector tile data in iOS. I guess I'll look at the mvt spec and implement that in iOS directly. I'm not looking to build a feature complete renderer, just enough to render our custom dataset.
